File: solutions/303-range-sum-query-immutable/range-sum-query-immutable.ts
class NumArray {

    prefixSums: number[];

    constructor(nums: number[]) {
        
        this.prefixSums = [];
        let sum = 0;
        for(let i = 0; i < nums.length; i+= 1) {
            sum += nums[i];
            this.prefixSums.push(sum);
        }

    }

    sumRange(left: number, right: number): number {
        
        return this.prefixSums[right] - (left > 0 ? this.prefixSums[left - 1] : 0);

    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: solutions/303-range-sum-query-immutable/range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums: list[int]):
        self.prefix_sums = [0]
        
        current_running_sum = 0
        for num in nums:
            current_running_sum += num
            self.prefix_sums.append(current_running_sum)

    def sumRange(self, left: int, right: int) -> int:
        return self.prefix_sums[right + 1] - self.prefix_sums[left]

        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)